### PR TITLE
DO NOT MERGE: move Network.default() to Network()

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
This replaces the `Network.default()` classmethod with defaults for the class, so you get the same object with just `Network`.

This is a placeholder PR to consider whether this is a good change or not - if we do want to go ahead with it, I'll open a new one against upstream/7.0 (for merging after the require-keywords PR is merged).

### CODE REVIEW

**Comment by tonyandrewmeyer** [comment]
@benhoyt we discussed this change a couple of times, most recently in Madrid. What do you think? The test_networks file in particular is one to look at, as it shows using `Network()` with defaults but changing one of the nested values.

I still don't really like `.default` as a name for the classmethod, and I think it's more "Scenario-like" to have reasonable default values for the state objects. It's not great that it's more complicated to get a `Network` that just has the `Address.value` different.

An alternative to this implementation would be to push the defaults lower, e.g. (the factories would actually be `__init__` code given the keyword PR, but for ease of reading I've used the 3.10+ approach here):

```python
@dataclasses.dataclass(frozen=True)
class Address:
    hostname: str = ""
    value: str = "192.0.2.0"
    cidr: str = ""

@dataclasses.dataclass(frozen=True)
class BindAddress:
    interface_name: str = ""
    addresses: List[Address] = dataclasses.field(default_factory=lambda: [Address()])
    mac_address: Optional[str] = None

@dataclasses.dataclass(frozen=True)
class Network:
    bind_addresses: List[BindAddress] = dataclasses.field(default_factory(lambda: [BindAddress()])
    ingress_addresses: List[str] = dataclasses.field(default_factory=lambda: ["192.0.2.0"])
    egress_subnets: List[str] = dataclasses.field(default_factory=lambda: ["192.0.2.0/24"])
```

It's roughly as much work to create a `Network` where the `Address` is different, although less if it's the `interface_name` or `mac_address` that you want different that the defaults.

This does feel a bit cleaner to me - but networks/spaces are not something I have done much with, so I could be missing some reason this is a bad idea.

Thoughts?

**Comment by benhoyt** [comment]
A couple of thoughts:

* It seems strange to me that `Address()` itself has a default for `value`. I think it's great that `Network()` has good defaults`, but maybe `Address()` shouldn't.
* I'm not sure whether that should apply to `BindAddress` or not. Probably the same there -- `Network` should have the clever/fixed defaults, but the others maybe not.
* Could `Address.value` by the first and positional argument? I think writing `Address("1.2.3.4")` is clear and useful.

So what about this?

```python
@dataclasses.dataclass(frozen=True)
class Address:
    value: str
    _: dataclasses.KW_ONLY
    hostname: str = ""
    cidr: str = ""

@dataclasses.dataclass(frozen=True)
class BindAddress:
    addresses: List[Address]
    _: dataclasses.KW_ONLY
    interface_name: str = ""
    mac_address: Optional[str] = None

@dataclasses.dataclass(frozen=True)
class Network:
    bind_addresses: List[BindAddress] = dataclasses.field(default_factory(lambda: [BindAddress([Address("192.0.2.0")])])
    ingress_addresses: List[str] = dataclasses.field(default_factory=lambda: ["192.0.2.0"])
    egress_subnets: List[str] = dataclasses.field(default_factory=lambda: ["192.0.2.0/24"])
```

File: tests/test_e2e/test_network.py [hunk 1]
```diff
@@ -40,6 +47,14 @@ def test_ip_get(mycharm):
...
     )
 
+    network = Network(
+        bind_addresses=[
+            BindAddress(
+                interface_name="",
+                addresses=[Address(hostname="", value="203.0.113.0", cidr="")],
+            ),
+        ],
+    )
```
Comment (benhoyt) [inline tests/test_e2e/test_network.py]: It seems like we should make use of the defaults here (even if just as a "good example"). And if `value` becomes required+positional, like my other comment, we'd have:

```python
network = Network(bind_addresses=BindAddress([Address("203.0.113.0")]))
```

Thoughts?

**Comment by tonyandrewmeyer** [comment]
> * I'm not sure whether that should apply to `BindAddress` or not. Probably the same there -- `Network` should have the clever/fixed defaults, but the others maybe not.

Ok, so then it should be the way it is in the PR and not the alternative in the comment, then. That works for me.

> * Could `Address.value` by the first and positional argument? I think writing `Address("1.2.3.4")` is clear and useful.

That sounds reasonable to me. To avoid conflicts, I'll do that in #137. Same for `BindAddress.addresses` and `BindAddress.interface`.

I'll update the branch to make use of the simpler implementation, and then open a PR against the upstream 7.0 once the kwargs branch is merged.

Thanks!

**Comment by tonyandrewmeyer** [comment]
Opened an upstream PR.

**Pull request closed without merging**